activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -29,7 +29,14 @@ def scope
         end
 
         def records_for(ids)
-          query_scope(ids)
+          records = query_scope(ids)
+          # Respect the order on `scope` if it exists, else use the order `ids`,
+          # which is already sorted based on the order of the through association
+          if scope.values[:order].any?
+            records
+          else
+            records.sort_by { |record| ids.index record[association_key_name] }
```

Comment:
Can we do this at the database level, instead? This will slow down reading through associations quite a bit. I think it would make more sense to implement this by using the order from the other scope if it's undefined on the through association.